File: backend-node/src/services/payment-fulfillment.service.js
import { getOrCreateUser, saveUser } from '../models/user.model.js';
import { createPurchase, registerCouponUse, getPurchaseByPaymentId } from '../services/supabase.service.js';
import { sendPurchaseConfirmationEmail } from '../services/email.service.js';

/**
 * Conclui pedido pago com idempotência por id_pagamento
 */
export const fulfillPaidOrder = async ({
  userId,
  planId,
  planName,
  analyses,
  price,
  paymentMethod,
  paymentId,
  customerEmail = '',
  couponId = null,
  couponName = null,
  discountPercent = null,
  originalPrice = null,
  cpfNormalized = null,
  extraInfo = ''
}) => {
  const existing = await getPurchaseByPaymentId(paymentId);
  if (existing) {
    const user = await getOrCreateUser(userId, customerEmail);
    return {
      alreadyFulfilled: true,
      user: {
        id: user.id,
        credits: user.credits,
        plan: user.plan
      }
    };
  }

  await createPurchase(
    userId,
    planId,
    planName,
    analyses,
    price,
    'BRL',
    paymentMethod,
    paymentId,
    null,
    'analysis_plan',
    couponId,
    couponName,
    discountPercent,
    originalPrice
  );

  if (couponId && cpfNormalized) {
    await registerCouponUse(couponId, cpfNormalized);
  }

  const user = await getOrCreateUser(userId, customerEmail);
  await user.addCredits(analyses);
  const updatedUser = await saveUser({ ...user, plan: planId });

  const email = customerEmail || updatedUser.email || '';
  if (email) {
    try {
      await sendPurchaseConfirmationEmail(email, {
        planName,
        analyses,
        price,
        customerName: updatedUser.name || '',
        extraInfo,
        couponName: couponName || undefined,
        discountPercent: discountPercent ?? undefined,
        originalPrice: originalPrice ?? undefined
      });
    } catch (emailErr) {
      console.error('Erro ao enviar confirmação de compra:', emailErr);
    }
  }

  return {
    alreadyFulfilled: false,
    user: {
      id: updatedUser.id,
      credits: updatedUser.credits,
      plan: updatedUser.plan
    }
  };
};

/**
 * Checkout 100% grátis (cupom)
 */
export const fulfillFreeCheckout = async (ctx, email = '') => {
  return fulfillPaidOrder({
    userId: ctx.userId,
    planId: ctx.planId,
    planName: ctx.planName,
    analyses: ctx.analyses,
    price: 0,
    paymentMethod: 'coupon',
    paymentId: `free_${Date.now()}_${ctx.userId}`,
    customerEmail: email,
    couponId: ctx.couponId,
    couponName: ctx.couponName,
    discountPercent: ctx.discountPercent,
    originalPrice: ctx.originalPrice,
    cpfNormalized: ctx.cpfNormalized,
    extraInfo: 'Compra 100% grátis com cupom.'
  });
};

File: backend/src/CurriculosProIA.Service/Implementations/PaymentFulfillmentService.cs
using CurriculosProIA.Domain.Dtos;
using CurriculosProIA.Repository.Interfaces;
using CurriculosProIA.Service.Interfaces;
using Microsoft.Extensions.Logging;

namespace CurriculosProIA.Service.Implementations;

public class PaymentFulfillmentService : IPaymentFulfillmentService
{
    private readonly IPurchaseRepository _purchases;
    private readonly IUserProfileRepository _users;
    private readonly ICouponRepository _coupons;
    private readonly ICreditRepository _credits;
    private readonly IAnalysisRepository _analyses;
    private readonly IPricingService _pricing;
    private readonly IEmailService _email;
    private readonly ILogger<PaymentFulfillmentService> _logger;

    public PaymentFulfillmentService(
        IPurchaseRepository purchases,
        IUserProfileRepository users,
        ICouponRepository coupons,
        ICreditRepository credits,
        IAnalysisRepository analyses,
        IPricingService pricing,
        IEmailService email,
        ILogger<PaymentFulfillmentService> logger)
    {
        _purchases = purchases;
        _users = users;
        _coupons = coupons;
        _credits = credits;
        _analyses = analyses;
        _pricing = pricing;
        _email = email;
        _logger = logger;
    }

    public Task<FulfillOrderResult> FulfillFreeCheckoutAsync(
        FulfillOrderRequest request,
        CancellationToken cancellationToken = default)
    {
        request.Price = 0;
        request.PaymentMethod = "coupon";
        request.PaymentId = $"free_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{request.UserId}";
        request.ExtraInfo = "Compra 100% grátis com cupom.";
        return FulfillPaidOrderAsync(request, cancellationToken);
    }

    public async Task<FulfillOrderResult> FulfillPaidOrderAsync(
        FulfillOrderRequest request,
        CancellationToken cancellationToken = default)
    {
        var existing = await _purchases.GetPurchaseByPaymentIdAsync(request.PaymentId, cancellationToken);
        if (existing != null)
        {
            var credits = await _credits.GetAvailableCreditsAsync(request.UserId, cancellationToken);
            var profile = await _users.GetUserProfileAsync(request.UserId, cancellationToken);
            return new FulfillOrderResult
            {
                AlreadyFulfilled = true,
                User = new PaymentUserSummary
                {
                    Id = request.UserId,
                    Credits = credits,
                    Plan = profile?.Plan
                }
            };
        }

        string? partnerId = null;
        decimal? partnerPercent = null;
        decimal? partnerAmount = null;

        if (!string.IsNullOrEmpty(request.CouponId))
        {
            var cupom = await _coupons.GetCouponByIdAsync(request.CouponId, cancellationToken);
            if (cupom != null && !string.IsNullOrEmpty(cupom.IdParceiro) && cupom.PorcentagemParceiro > 0)
            {
                partnerId = cupom.IdParceiro;
                partnerPercent = cupom.PorcentagemParceiro;
                partnerAmount = Math.Round(request.Price * cupom.PorcentagemParceiro.Value / 100m, 2);
            }
        }

        var isEnglishOnly = request.PlanId == "english";
        var config = await _pricing.GetPricingConfigAsync(cancellationToken);

        if (isEnglishOnly)
        {
            if (string.IsNullOrWhiteSpace(request.AnalysisId))
            {
                throw new InvalidOperationException("analysisId é obrigatório para compra de currículo em inglês.");
            }

            await _purchases.CreatePurchaseAsync(
                request.UserId,
                "english",
                "Currículo em Inglês",
                0,
                request.Price,
                request.PaymentMethod,
                request.PaymentId,
                serviceType: "curriculo_ingles",
                analysisId: request.AnalysisId,
                couponId: request.CouponId,
                couponName: request.CouponName,
                discountPercent: request.DiscountPercent,
                originalPrice: request.OriginalPrice,
                partnerId: partnerId,
                partnerPercent: partnerPercent,
                partnerAmount: partnerAmount,
                cancellationToken: cancellationToken);

            await _analyses.GrantEnglishPaidAsync(request.AnalysisId, cancellationToken);
        }
        else
        {
            var purchase = await _purchases.CreatePurchaseAsync(
                request.UserId,
                request.PlanId,
                request.PlanName,
                request.Analyses,
                request.Price,
                request.PaymentMethod,
                request.PaymentId,
                serviceType: "analysis_plan",
                couponId: request.CouponId,
                couponName: request.CouponName,
                discountPercent: request.DiscountPercent,
                originalPrice: request.OriginalPrice,
                partnerId: partnerId,
                partnerPercent: partnerPercent,
                partnerAmount: partnerAmount,
                cancellationToken: cancellationToken);

            if (request.IncludeEnglish)
            {
                var englishPrice = request.EnglishPriceBRL > 0
                    ? request.EnglishPriceBRL
                    : config.EnglishBundlePriceBRL;

                await _purchases.CreatePurchaseAsync(
                    request.UserId,
                    "english",
                    "Currículo em Inglês (bundle)",
                    0,
                    englishPrice,
                    request.PaymentMethod,
                    $"{request.PaymentId}_english",
                    parentPurchaseId: purchase.Id,
                    serviceType: "curriculo_ingles",
                    cancellationToken: cancellationToken);
            }
        }

        if (!string.IsNullOrEmpty(request.CouponId) && !string.IsNullOrEmpty(request.CpfNormalized))
        {
            await _coupons.RegisterCouponUseAsync(request.CouponId, request.CpfNormalized, cancellationToken);
        }

        if (!string.IsNullOrEmpty(request.CustomerEmail))
        {
            try
            {
                await _users.GetOrCreateUserProfileAsync(
                    request.UserId,
                    request.CustomerEmail,
                    cancellationToken: cancellationToken);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Erro ao garantir perfil do usuário após pagamento");
            }
        }

        if (!isEnglishOnly)
        {
            await _users.UpdateUserProfileAsync(
                request.UserId,
                new Dictionary<string, object?> { ["plan"] = request.PlanId },
                cancellationToken);
        }

        var creditsAvailable = await _credits.GetAvailableCreditsAsync(request.UserId, cancellationToken);
        var userProfile = await _users.GetUserProfileAsync(request.UserId, cancellationToken);

        var recipientEmail = !string.IsNullOrWhiteSpace(request.CustomerEmail)
            ? request.CustomerEmail.Trim()
            : userProfile?.Email?.Trim();

        if (!string.IsNullOrWhiteSpace(recipientEmail))
        {
            try
            {
                await _email.SendPurchaseConfirmationEmailAsync(
                    recipientEmail,
                    new PurchaseConfirmationDetails
                    {
                        PlanName = request.PlanName,
                        CreditsAmount = isEnglishOnly ? null : request.Analyses,
                        Price = request.Price,
                        CustomerName = userProfile?.Name,
                        CouponName = request.CouponName,
                        DiscountPercent = request.DiscountPercent,
                        OriginalPrice = request.OriginalPrice,
                        ExtraInfo = string.IsNullOrWhiteSpace(request.ExtraInfo) ? null : request.ExtraInfo
                    },
                    cancellationToken);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Erro ao enviar confirmação de compra para {Email}", recipientEmail);
            }
        }

        return new FulfillOrderResult
        {
            AlreadyFulfilled = false,
            User = new PaymentUserSummary
            {
                Id = request.UserId,
                Credits = creditsAvailable,
                Plan = userProfile?.Plan ?? request.PlanId
            }
        };
    }
}
